### DISCUSSION TITLE
UCXError while using RAPIDS Shuffle Manager

### DISCUSSION THREAD

**Discussion by Nikhilpa1** [NVIDIA/spark-rapids#8134] [Q&A]
While I try to run a spark job on GPU with RAPIDS Shuffle Manager 
UCX 1.14
Spark 3.2.0
rapids-4-spark_2.12-22.10.0
with configuration 
$SPARK_HOME/bin/spark-submit \
       --master spark://${MASTER_HOST}:7077 \
       --conf spark.executor.extraClassPath=${SPARK_RAPIDS_PLUGIN_JAR} \
       --conf spark.driver.extraClassPath=${SPARK_RAPIDS_PLUGIN_JAR} \
	--conf spark.plugins=com.nvidia.spark.SQLPlugin \
	--conf spark.shuffle.manager=com.nvidia.spark.rapids.spark320.RapidsShuffleManager \
       --conf spark.rapids.sql.concurrentGpuTasks=1 \
       --conf spark.driver.memory=64G \
       --conf spark.executor.memory=64G \
       --conf spark.executor.cores=1 \
	--conf spark.executor.instances=1 \
       --conf spark.task.cpus=1 \
       --conf spark.executor.resource.gpu.amount=1 \
       --conf spark.task.resource.gpu.amount=0.25 \
       --conf spark.rapids.memory.pinnedPool.size=1G \
       --conf spark.sql.files.maxPartitionBytes=128m \
	 --conf spark.rapids.shuffle.mode=UCX \
	 --conf spark.shuffle.service.enabled=false \
	 --conf spark.dynamicAllocation.enabled=false \
        --conf spark.executorEnv.LD_LIBRARY_PATH=/home/kanaka.3/others/ucx/ucx-1.14-ins/lib:/home/kanaka.3/others/knem/knem-1.1.4-ins/lib \
        --conf spark.driverEnv.LD_LIBRARY_PATH=/home/kanaka.3/others/ucx/ucx-1.14-ins/lib::/home/kanaka.3/others/knem/knem-1.1.4-ins/lib \
        --conf spark.executorEnv.UCX_ERROR_SIGNALS= \
	 --conf spark.executorEnv.UCX_MEMTYPE_CACHE=n \
	 --conf spark.executorEnv.UCX_IB_RX_QUEUE_LEN=1024 \
	--conf spark.executorEnv.UCX_TLS=rc_x,cuda_copy,cuda_ipc \
 	--conf spark.executorEnv.UCX_RNDV_SCHEME=put_zcopy \
 	--conf spark.executorEnv.UCX_MAX_RNDV_RAILS=1 \
${SPARK_JOBS}/myjob.py

I'm facing connection error

0/stderr:75:23/04/17 12:06:13 INFO UCXClientConnection: UCX Client UCXClientConnection(ucx=com.nvidia.spark.rapids.shuffle.ucx.UCX@100838be, peerExecutorId=1) started
0/stderr:76:23/04/17 12:06:13 ERROR UCX: UcpListener detected an error for executorId 1: UCXError(-6,Destination is unreachable)
0/stderr:77:23/04/17 12:06:13 WARN UCX: Removing endpoint UcpEndpoint(id=47937671962800, UcpEndpointParams{errorHandlingMode=UCP_ERR_HANDLING_MODE_PEER,socketAddress=/10.1.1.3:4195,) for 1
0/stderr:78:23/04/17 12:06:13 WARN UCX: Removed stale client connection for 1
0/stderr:79:23/04/17 12:06:13 ERROR UCX: Error while closing ep. Ignoring.
0/stderr:80:org.openucx.jucx.UcxException: Destination is unreachable
0/stderr:81:	at org.openucx.jucx.ucp.UcpEndpoint.closeNonBlockingNative(Native Method)
0/stderr:82:	at org.openucx.jucx.ucp.UcpEndpoint.closeNonBlockingFlush(UcpEndpoint.java:376)
0/stderr:83:	at com.nvidia.spark.rapids.shuffle.ucx.UCX$UcpEndpointManager.$anonfun$closeEndpointOnWorkerThread$1(UCX.scala:900)
0/stderr:84:	at com.nvidia.spark.rapids.shuffle.ucx.UCX.$anonfun$init$5(UCX.scala:184)
0/stderr:85:	at com.nvidia.spark.rapids.shuffle.ucx.UCX.$anonfun$init$5$adapted(UCX.scala:178)
0/stderr:86:	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)
0/stderr:87:	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)
0/stderr:88:	at com.nvidia.spark.rapids.shuffle.ucx.UCX.withResource(UCX.scala:69)
0/stderr:89:	at com.nvidia.spark.rapids.shuffle.ucx.UCX.$anonfun$init$2(UCX.scala:178)
0/stderr:90:	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
0/stderr:91:	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
0/stderr:92:	at com.nvidia.spark.rapids.GpuDeviceManager$$anon$1.$anonfun$newThread$1(GpuDeviceManager.scala:345)
0/stderr:93:	at java.lang.Thread.run(Thread.java:748)

I've checked on UCX communication by validating the nodes with ucx_perftest. (UCX_NET_DEVICES=mlx5_0:1 UCX_SOCKADDR_TLS_PRIORITY=sockcm UCX_TLS=rc_x,cuda_copy,cuda_ipc ./ucx_perftest -t tag_bw -m cuda -n 4000 -w 500 -c 0 -s 134217728 -p 13381)

**Comment by abellina** [comment]
@Nikhilpa1, you passed: `UCX_NET_DEVICES=mlx5_0:1 UCX_SOCKADDR_TLS_PRIORITY=sockcm` to ucx_perftest, but I don't see either of these variables passed to the spark job. What happens if you set all the environment variables as your ucx_perftest example?

**Reply by Nikhilpa1** [reply]
When I include `UCX_NET_DEVICES=mlx5_0:1 UCX_SOCKADDR_TLS_PRIORITY=sockcm`

--conf spark.executorEnv.UCX_NET_DEVICES=mlx5_0:1 \
--conf spark.executorEnv.UCX_SOCKADDR_TLS_PRIORITY=sockcm \

It throws,
0/stderr:68:23/04/17 12:26:58 INFO UCX: UCX Worker created
0/stderr:69:23/04/17 12:26:58 WARN RapidsConf: CUDA drivers before 11.5 have known incompatibilities with the ASYNC allocator, falling back to ARENA
0/stderr:70:23/04/17 12:26:58 ERROR RapidsShuffleHeartbeatEndpoint: Error initializing shuffle
0/stderr:71:java.net.BindException: **Couldn't start UcpListener on port range 26957-26973**
0/stderr:72:	at com.nvidia.spark.rapids.shuffle.ucx.UCX$UcpEndpointManager.startListener(UCX.scala:981)
0/stderr:73:	at com.nvidia.spark.rapids.shuffle.ucx.UCX.startListener(UCX.scala:211)
0/stderr:74:	at com.nvidia.spark.rapids.shuffle.ucx.UCXServerConnection.startManagementPort(UCXConnection.scala:51)
0/stderr:75:	at com.nvidia.spark.rapids.shuffle.RapidsShuffleServer.start(RapidsShuffleServer.scala:131)
0/stderr:76:	at org.apache.spark.sql.rapids.RapidsShuffleInternalManagerBase.server$lzycompute(RapidsShuffleInternalManagerBase.scala:1136)
0/stderr:77:	at org.apache.spark.sql.rapids.RapidsShuffleInternalManagerBase.server(RapidsShuffleInternalManagerBase.scala:1120)
0/stderr:78:	at org.apache.spark.sql.rapids.RapidsShuffleInternalManagerBase.getServerId(RapidsShuffleInternalManagerBase.scala:1019)

**Reply by abellina** [reply]
OK thanks for trying that. It looks like there are other differences between the two (ucx_perftest vs. the app), and I am wondering if it's the executor's user versus the ucx_perftest user.

The last error is UCX trying to startup a TCP listener and failing to bind to the socket after 16 attempts.

I think it might be best to remove configs and try adding them one by one. My guess is that the UCX_TLS selection is causing some of it. 

Could you run ucx_perftest without _any_ environment variables? If that works, it would be good to try the spark job with:

```
--conf spark.executorEnv.UCX_ERROR_SIGNALS=
--conf spark.executorEnv.UCX_MEMTYPE_CACHE=n
--conf spark.executorEnv.UCX_IB_RX_QUEUE_LEN=1024
```

UCX_LOG_LEVEL=data and UCX_LOG_FILE can be used to capture logs (it would be good to do this for both ucx_perftest and the spark executor). If you still run into trouble, we can get someone from the UCX team to also take a look.

**Reply by Nikhilpa1** [reply]
Thanks for the reply. I can't run ucx_perftest successfully without environment variables as when I do,

./ucx_perftest -t tag_bw -m cuda -n 4000 -w 500 -c 0 -s 134217728
Waiting for connection...
Accepted connection from 10.1.1.3:41952
+----------------------------------------------------------------------------------------------------------+
| API:          protocol layer                                                                             |
| Test:         tag match bandwidth                                                                        |
| Data layout:  (automatic)                                                                                |
| Send memory:  cuda                                                                                       |
| Recv memory:  cuda                                                                                       |
| Message size: 134217728                                                                                  |
+----------------------------------------------------------------------------------------------------------+
[1681751926.324058] [gpu02:30081:0]          wireup.c:1205 UCX  ERROR   old: am_lane 0 wireup_msg_lane 1 cm_lane <none> keepalive_lane <none> reachable_mds 0x42
[1681751926.324085] [gpu02:30081:0]          wireup.c:1215 UCX  ERROR   old: lane[0]: 10:rc_mlx5/mlx5_0:1.0 md[6]      -> md[6]/ib/sysdev[2] rma_bw#0 am am_bw#0
[1681751926.324090] [gpu02:30081:0]          wireup.c:1215 UCX  ERROR   old: lane[1]:  2:tcp/en0.0 md[1]               -> md[1]/tcp/sysdev[2] rma_bw#1 wireup
[1681751926.324094] [gpu02:30081:0]          wireup.c:1205 UCX  ERROR   new: am_lane 0 wireup_msg_lane 1 cm_lane <none> keepalive_lane <none> reachable_mds 0x42
[1681751926.324098] [gpu02:30081:0]          wireup.c:1215 UCX  ERROR   new: lane[0]: 10:rc_mlx5/mlx5_0:1.0 md[6]      -> md[6]/ib/sysdev[2] rma_bw#0 am am_bw#0
[1681751926.324102] [gpu02:30081:0]          wireup.c:1215 UCX  ERROR   new: lane[1]:  2:tcp/en0.0 md[1]               -> md[1]/tcp/sysdev[1] rma_bw#1 wireup
[gpu02:30081:0:30081]      wireup.c:1554 Fatal: endpoint reconfiguration not supported yet

/home/kanaka.3/others/ucx/ucx-1.14.0/src/ucp/wireup/wireup.c: [ ucp_wireup_init_lanes() ]
      ...
     1551                                 NULL, cm_idx, UCS_LOG_LEVEL_ERROR);
     1552         ucp_wireup_print_config(worker, &key, "new", NULL,
     1553                                 cm_idx, UCS_LOG_LEVEL_ERROR);
==>  1554         ucs_fatal("endpoint reconfiguration not supported yet");
     1555     }
     1556
     1557     ep->cfg_index = new_cfg_index;

**Reply by abellina** [reply]
@Nikhilpa1 mind sharing `ucx_info -d` ? It should tell us what devices and transports are available. Also, adding @yosefe and @petro-rudenko to the thread in case they have insights.

**Reply by Nikhilpa1** [reply]
Sure. Please find it here,

./ucx_info -d | grep "Device\|Transport:"
      Transport: self
         Device: memory
      Transport: tcp
         Device: lo
      Transport: tcp
         Device: en0
      Transport: tcp
         Device: ib0
      Transport: sysv
         Device: memory
      Transport: posix
         Device: memory
      Transport: cuda_copy
         Device: cuda
      Transport: cuda_ipc
         Device: cuda
      Transport: dc_mlx5
         Device: mlx5_0:1
      Transport: rc_verbs
         Device: mlx5_0:1
      Transport: rc_mlx5
         Device: mlx5_0:1
      Transport: ud_verbs
         Device: mlx5_0:1
      Transport: ud_mlx5
         Device: mlx5_0:1
      Transport: cma
         Device: memory
      Transport: knem
         Device: memory

**Reply by abellina** [reply]
Thanks @Nikhilpa1. Can we try: `UCX_NET_DEVICES=en0 UCX_TLS=tcp,cuda_copy,cuda_ipc` for both ucx_perftest and the spark-submit?

Assuming that `en0` interface is available everywhere that spark runs and that all executors can connect between themselves via `en0`. 

This is not a recommended config, but I think it might be the simplest just to get us going.

**Reply by Nikhilpa1** [reply]
`UCX_NET_DEVICES=en0 UCX_TLS=tcp,cuda_copy,cuda_ipc` seems to work but is more likely to perform poorly than when UCX is turned OFF.

I'm trying to run TPC-DS benchmarks with and without UCX and compare the performance.
FYI, Configuration for UCX OFF:
export SPARK_CONF=(
    "--master" "spark://${MASTER_HOST}:7077"
    "--conf" "spark.executor.extraClassPath=${SPARK_RAPIDS_PLUGIN_JAR}"
    "--conf" "spark.driver.extraClassPath=${SPARK_RAPIDS_PLUGIN_JAR}"
    "--conf" "spark.plugins=com.nvidia.spark.SQLPlugin"
    "--conf" "spark.rapids.sql.concurrentGpuTasks=1"
    "--conf" "spark.driver.memory=64G"
    "--conf" "spark.executor.memory=64G"
    "--conf" "spark.executor.cores=1"
    "--conf" "spark.executor.instances=1"
    "--conf" "spark.task.cpus=1"
    "--conf" "spark.executor.resource.gpu.amount=1"
    "--conf" "spark.task.resource.gpu.amount=0.25"
    "--conf" "spark.rapids.memory.pinnedPool.size=1G"
    "--conf" "spark.sql.files.maxPartitionBytes=128m"
    "--conf" "spark.shuffle.service.enabled=false"
    "--conf" "spark.sql.shuffle.partitions=200"
    "--conf" "spark.rapids.sql.hasNans=false"
    "--conf" "spark.rapids.sql.incompat=true"
    "--conf" "spark.rapids.sql.variableFloatAgg.enabled=true"
)

**Reply by abellina** [reply]
Yes, performance won't be great with just TCP, but this helps triage a bit. If this is working with ucx_perftest and our plugin, it means:

- Executors can talk to each other using TCP (which is the type of error we were getting before with the bind error)
- The cuda runtime is visible to UCX (ucx_info -d showed that already), but you can now see the UCX pipelined protocol working here (cuda_copy). 

Here are some questions:

1. Are the machines used by the executors the same as the machines you used for ucx_perftest?
2. Are you using RoCE or Infiniband? One thing that stands out is the `ib0` interface, and I believe that is normally "Ethernet-over-infiniband". I wonder if things are routing through this interface in some scenarios, and in some other scenarios via en0.

Some next steps:
1. Lets add to UCX_NET_DEVICES the Mellanox device: `UCX_NET_DEVICES=en0,mlx5_0:1`.
2. If the above works, stop specifying `UCX_TLS`, so it picks defaults => then Infiniband/RoCE should be active. 

I am also curious on what happens if ib0 is added to the mix, with or without the UCX_TLS that worked (tcp,cuda_ipc,cuda_copy). 

Hope we get closer, if you are still having trouble, sharing the UCX log as an attachment is going to be my next ask so we can discuss with the UCX team.

**Reply by Nikhilpa1** [reply]
I agree with the inference. To further answer your questions,
1. Yes, the machines used by the executors are the same as the machines I used for ucx_perftest.
2. Infiniband is being used in the system.

I've tried the suggestions,
1. Performance improvement was observed when the Mellanox device is added without UCX_TLS but faced disconnections 
``0/stderr:78:23/04/17 17:43:55 ERROR UCX: UcpListener detected an error for executorId 1: UCXError(-25,Connection reset by remote peer)``

**Reply by abellina** [reply]
Thanks @Nikhilpa1, I have one more thing to try, but please be aware we are debugging what looks to be a similar issue than yours (https://github.com/openucx/ucx/issues/8971) right now, so I should have more info shortly.

One thing to try is to turn off our wakeup/epoll setting. We ask UCX to send its thread to sleep instead of having a hot loop for its progress loop. I would like to know if turning off wakeup helps your case:

`--conf spark.rapids.shuffle.ucx.useWakeup=false`

Thanks for trying all of the configs so far.
